examples/cis2-wccd-upgradable/src/lib.rs

```diff
@@ -0,0 +1,3721 @@
+    _ctx: &'b impl HasReceiveContext,
+    host: &'a impl HasHost<StateImplementation, StateApiType = S>,
+) -> ContractResult<&'a StateImplementation> {
+    Ok(host.state())
+}
+
+/// Function to view state of contract.
+#[receive(contract = "CIS2-wCCD-Proxy", name = "view", return_value = "StateProxy")]
+fn contract_proxy_view<'a, 'b, S: HasStateApi>(
+    _ctx: &'b impl HasReceiveContext,
+    host: &'a impl HasHost<StateProxy, StateApiType = S>,
+) -> ContractResult<&'a StateProxy> {
+    Ok(host.state())
+}
+
+/// Helper function to ensure contract is not paused.
+fn when_not_paused<S>(
+    host: &mut impl HasHost<StateImplementation, StateApiType = S>,
+) -> ContractResult<()> {
+    let state_address = if let ProtocolAddressesImplementation::Initialized {
+        proxy_address: _,
+        state_address,
+    } = host.state().protocol_addresses
+    {
+        state_address
+    } else {
+        panic!("{:?}", concordium_cis2::Cis2Error::Custom(CustomContractError::UnInitialized));
+    };
+
+    let paused = host.invoke_contract_raw_read_only(
+        &state_address,
+        Parameter(&[]),
+        EntrypointName::new_unchecked("getPaused"),
+        Amount::zero(),
+    )?;
+
+    // It is expected that this contract is initialized with the w_ccd_state
+    // contract (a V1 contract). In that case, the paused variable can be
+    // queried from the state contract without error.
+    let paused: bool = paused
+        .ok_or(concordium_cis2::Cis2Error::Custom(CustomContractError::StateInvokeError))?
+        .get()?;
+    // Check that contract is not paused.
+    ensure!(!paused, concordium_cis2::Cis2Error::Custom(CustomContractError::ContractPaused));
+    Ok(())
+}
+
+/// Wrap an amount of CCD into wCCD tokens and transfer the tokens if the sender
+/// is not the receiver.
+#[receive(contract = "CIS2-wCCD", name = "wrap", mutable, payable)]
+fn contract_wrap<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<StateImplementation, StateApiType = S>,
+    amount: Amount,
+) -> ContractResult<()> {
+    let (proxy_address, state_address) = if let ProtocolAddressesImplementation::Initialized {
+        proxy_address,
+        state_address,
+    } = host.state().protocol_addresses
+    {
+        (proxy_address, state_address)
+    } else {
+        panic!("{:?}", concordium_cis2::Cis2Error::Custom(CustomContractError::UnInitialized));
+    };
+
+    // Can be only called through the fallback function on the proxy.
+    only_proxy(proxy_address, ctx.sender())?;
+
+    // Check that contract is not paused.
+    when_not_paused(host)?;
+    if amount == Amount::zero() {
+        return Ok(());
+    }
+
+    let input: WrapParamsWithSender = ctx.parameter_cursor().get()?;
+
+    // Get the sender who invoked this contract function.
+    let sender = input.sender;
+
+    let receive_address = input.params.to.address();
+
+    // Proxy holds CCD funds. CCD is sent to proxy.
+    host.invoke_contract_raw(
```

Comment:
Instead of moving the CCD back and forth between proxy and implementation, it might be enough to just pass the information back removing the need to call back the proxy with the CCD.

Reply:
The CCD amount is passed by the fallback function on the proxy to the implementation. The implementation has the logic to check how much WCCD should be minted (based on the amount of CCD it receives in the `wrap` function call). This logic has to be executed before the WCCD can be sent to the proxy (sent to be stored on the proxy). The CCD should be stored on the proxy in case we upgrade the implementation.

Can you please explain how you can pass the information back or improve the pattern based on your above comment?

Reply:
I was thinking adding the amount of tokens (just like the sender address), such that the information of how much was transferred is forwarded to the implementation, but not the actual CCDs.
But I am not sure whether it will simplify anything.

Reply:
Since the fallback function will be used by all state-mutative functions, we would also append the "amount of tokens" to the "transfer", "unwrap", and "updateOperator" functions which do not need that information. I am not sure what is the best (most efficient/ simplest) way of handling it.